File: src/slidr/render/html.py
"""HTML renderer for slidr."""

import os
import subprocess
import tempfile
from pathlib import Path
from jinja2 import Environment, FileSystemLoader
from pygments import highlight
from pygments.lexers import get_lexer_by_name, TextLexer
from pygments.formatters import HtmlFormatter
from pygments.util import ClassNotFound

from slidr.parser.ast import Document
from slidr.render.ir import build_ir, SlideIR, Elem
from slidr.plugins.layouts import KNOWN_LAYOUTS

_pygments_style = "default"

from slidr.render.ir import build_ir, SlideIR, Elem

_BRAND_ICON_SVGS = {
    "linkedin": (
        '<svg class="speaker-icon" xmlns="http://www.w3.org/2000/svg" '
        'viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" '
        'stroke-linecap="round" stroke-linejoin="round" '
        'style="height:1em;width:auto;vertical-align:middle">'
        '<path d="M16 8a6 6 0 0 1 6 6v7h-4v-7a2 2 0 0 0-2-2 2 2 0 0 0-2 2v7h-4v-7a6 6 0 0 1 6-6z"/>'
        '<rect width="4" height="12" x="2" y="9"/>'
        '<circle cx="4" cy="4" r="2"/></svg>'
    ),
}

TEMPLATE_DIR = Path(__file__).parent / "templates"
THEME_DIR = Path(__file__).parent.parent / "themes"
_env = Environment(loader=FileSystemLoader(str(TEMPLATE_DIR)), autoescape=False)


def default_theme() -> str:
    p = THEME_DIR / "default.css"
    return p.read_text() if p.exists() else ""


def load_theme(name: str) -> str:
    """Load a named theme CSS file. Falls back to default if not found."""
    p = THEME_DIR / f"{name}.css"
    return p.read_text() if p.exists() else default_theme()


def base_css() -> str:
    p = TEMPLATE_DIR / "base.css"
    return p.read_text() if p.exists() else ""


def render(doc: Document, theme_css: str, assembled_css: str, dims: tuple[int, int], logo: str = "") -> str:
    from slidr.render.seaborn import set_palette

    global _pygments_style
    _pygments_style = doc.meta.pygments_style or "default"
    set_palette(doc.meta.seaborn_theme)

    ir_slides = build_ir(doc, base_css(), theme_css)

    slides = []
    for i, slide in enumerate(ir_slides):
        children = _render_slide(slide)
        slides.append({
            "num": i + 1, "layout": slide.layout, "children": children,
            "notes": slide.notes, "footer": doc.meta.footer or "", "paginate": doc.meta.paginate or False,
            "variant": slide.variant or doc.meta.theme_variant,
            "transition": slide.transition or doc.meta.transition or "",
        })

    return _env.get_template("shell.html").render(
        title=doc.meta.title or "Presentation",
        slide_w=dims[0], slide_h=dims[1],
        css=assembled_css, slides=slides,
        variant=doc.meta.theme_variant or "light",
    )


def _render_slide(slide: SlideIR) -> str:
    elems = slide.elements
    heading_html = ""
    if elems and elems[0].kind == "heading":
        heading_html = _render_elem(elems[0])
        elems = elems[1:]
    if elems and elems[0].kind == "subtitle":
        heading_html += "\n" + _render_elem(elems[0])
        elems = elems[1:]

    body_html = "\n".join(filter(None, (_render_elem(e) for e in elems)))
    if body_html.strip():
        heading_html += '\n<div class="slide-body">\n' + body_html + '\n</div>'
    return heading_html or body_html


def _render_elem(e: Elem) -> str:
    if e.kind == "heading":
        tag = f"h{e.level}"
        return f"<{tag}>{e.content}</{tag}>"
    elif e.kind == "text":
        return f"<p>{e.content}</p>" if e.content else ""
    elif e.kind == "quote":
        return f'<div class="quote">{e.content}</div>' if e.content else ""
    elif e.kind == "code":
        if e.language == "mermaid":
            return _render_mermaid(e.content, e.svg)
        if e.language == "seaborn":
            return _render_seaborn_html(e.content, e.svg)
        if e.language == "dot":
            return _render_dot_html(e.content, e.svg)
        return _highlight_code(e.content, e.language)
    elif e.kind == "list":
        s = "<ul>\n"
        for item in e.items:
            s += f"<li>{item}</li>\n"
        s += "</ul>"
        return s
    elif e.kind == "table":
        s = "<table>\n<thead>\n<tr>"
        for h in e.headers:
            s += f"<th>{_maybe_escape(h)}</th>"
        s += "</tr>\n</thead>\n<tbody>\n"
        for row in e.rows:
            s += "<tr>"
            for cell in row:
                s += f"<td>{_maybe_escape(cell)}</td>"
            s += "</tr>\n"
        s += "</tbody>\n</table>"
        return s
    elif e.kind == "grid":
        if e.layout:
            cls = f"{e.layout}"
            if "compare" in cls:
                # children: col-left, arrow, col-right
                parts = "\n".join(filter(None, (_render_elem(c) for c in e.children)))
                return f'<div class="{cls}">\n{parts}\n</div>'
            children = "\n".join(filter(None, (_render_elem(c) for c in e.children)))
            return f'<div class="{cls}">\n{children}\n</div>'
        cols = e.cols or len(e.children) or 2
        style = f"grid-template-columns: repeat({cols}, 1fr); gap: 16px;"
        cls = "grid"
        if e.class_:
            cls += f" {e.class_}"
        children = "\n".join(filter(None, (_render_elem(c) for c in e.children)))
        return f'<div class="{cls}" style="{style}">\n{children}\n</div>'
    elif e.kind == "card":
        cls = "card"
        if e.tag:
            cls += f" tag-{e.tag}"
        if e.class_:
            cls += f" {e.class_}"
        s = f'<div class="{cls}">\n'
        if e.header:
            s += f"<h3>{_maybe_escape(e.header)}</h3>\n"
        for line in e.body:
            line = _maybe_escape(line)
            if line.startswith("<ul") or line.startswith("<ol"):
                s += line + "\n"
            else:
                s += f"<p>{line}</p>\n"
        for child in e.children:
            s += _render_elem(child) + "\n"
        s += "</div>"
        return s
    elif e.kind == "arrow":
        return f'<div class="card-arrow">{e.content}</div>'
    elif e.kind == "row":
        cols = len(e.children)
        children = "\n".join(filter(None, (_render_elem(c) for c in e.children)))
        return f'<div class="row" style="display:grid;grid-template-columns:repeat({cols},1fr);gap:1em">\n{children}\n</div>'
    elif e.kind == "notes":
        cls = "notes"
        if e.tag:
            cls += f" tag-{e.tag}"
        return f'<div class="{cls}">{e.content}</div>'
    elif e.kind == "speaker":
        from slidr.plugins.lucide import render_icon
        name = e.attrs.get("name", e.content)
        role = e.attrs.get("role", "")
        text = f"{_escape(name)}<br><span class=\"role\">{_escape(role)}</span>" if role else _escape(name)
        links = []
        contact_icons = {"github": "git-fork", "twitter": "bird", "email": "mail", "website": "globe"}
        for key in ("github", "twitter", "email", "linkedin", "website"):
            val = e.attrs.get(key, "")
            if val:
                icon = contact_icons.get(key, "link")
                href = val if "://" in val else f"https://{val}" if key != "email" else f"mailto:{val}"
                label = val.replace("https://", "").replace("mailto:", "")
                svg = _BRAND_ICON_SVGS.get(key) or render_icon(icon, {"cls": "speaker-icon"})
                links.append(f'<a href="{_escape(href)}" target="_blank">{svg} {_escape(label)}</a>')
        if links:
            text += '<br><span class="speaker-links">' + "".join(links) + "</span>"
        return f'<div class="speaker">{text}</div>'
    elif e.kind in ("kicker", "subtitle", "tiny"):
        return f'<p class="{e.kind}">{e.content}</p>'
    return ""


def _highlight_code(content: str, language: str) -> str:
    try:
        lexer = get_lexer_by_name(language, stripall=True) if language else TextLexer()
    except ClassNotFound:
        lexer = TextLexer()
    formatter = HtmlFormatter(nowrap=True, style=_pygments_style)
    highlighted = highlight(content, lexer, formatter)
    cls = f' class="highlight language-{language}"' if language else ' class="highlight"'
    return f'<pre{cls}><code>{highlighted}</code></pre>'


def _pygments_css(style: str = "default") -> str:
    formatter = HtmlFormatter(style=style)
    css = formatter.get_style_defs('.slide .highlight')
    return f"\n/* ---- pygments ---- */\n{css}\n"


def _render_seaborn_html(content: str, svg: str = "") -> str:
    if svg:
        return f'<div class="seaborn-plot">\n{svg}\n</div>'
    return f'<pre class="seaborn-fallback"><code>{_escape(content)}</code></pre>'


def _render_mermaid(content: str, svg: str = "") -> str:
    if svg:
        return f'<div class="mermaid">\n{svg}\n</div>'
    return f'<pre class="mermaid-fallback"><code>{_escape(content)}</code></pre>'


def _render_dot_html(content: str, svg: str = "") -> str:
    if svg:
        return f'<div class="graphviz-plot">\n{svg}\n</div>'
    return f'<pre class="dot-fallback"><code>{_escape(content)}</code></pre>'


def _normalize_viewbox(svg: str) -> str:
    import re
    m = re.search(r'viewBox="([\d.-]+)\s+([\d.-]+)\s+([\d.]+)\s+([\d.]+)"', svg)
    if not m:
        return svg
    x, y, w, h = float(m.group(1)), float(m.group(2)), float(m.group(3)), float(m.group(4))
    if x >= 0 and y >= 0:
        return svg
    new_w, new_h = w + abs(x), h + abs(y)
    return svg.replace(m.group(0), f'viewBox="0 0 {new_w:g} {new_h:g}"', 1)


def _escape(s: str) -> str:
    return s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace('"', "&quot;")


def _maybe_escape(s: str) -> str:
    """Escape HTML unless the string already contains HTML tags."""
    return s if "<" in s and ">" in s else _escape(s)

File: src/render/html.rs
use crate::parser::ast::*;
use std::fmt::Write;

pub fn render(doc: &Document, theme_css: &str, logo: Option<&str>) -> String {
    let dims = doc.meta.dimensions();
    let ratio = dims.0 as f64 / dims.1 as f64;

    let mut slides_html = String::new();
    for (i, slide) in doc.slides.iter().enumerate() {
        let num = i + 1;
        let layout = match slide.layout {
            LayoutType::Title => "title",
            LayoutType::Grid2 => "grid-2",
            LayoutType::Grid3 => "grid-3",
            LayoutType::Grid4 => "grid-4",
            LayoutType::Content => "content",
        };
        let children = slide.children.iter()
            .filter_map(|n| render_node(n))
            .collect::<Vec<_>>()
            .join("\n");

        let footer = doc.meta.footer.as_deref().unwrap_or("");
        let paginate = doc.meta.paginate.unwrap_or(false);
        let mut footer_html = String::new();
        if !footer.is_empty() {
            write!(footer_html, "<footer>{}{}</footer>",
                footer,
                if paginate { format!(" &mdash; {}", num) } else { String::new() }
            ).unwrap();
        }

        write!(slides_html, r#"<section class="slide layout-{}"{}>
{}{}
</section>"#,
            layout,
            if slide.notes.is_empty() { String::new() } else { format!(r#" data-notes="{}""#, slide.notes) },
            children,
            footer_html,
        ).unwrap();
    }

    let logo_css = logo.map(|l| format!(
        r#"section::before {{
  content: "";
  position: absolute;
  top: 4%;
  right: 5%;
  width: 14%;
  height: 0;
  padding-bottom: 6%;
  background: url("{}") center / contain no-repeat;
  opacity: 0.92;
}}"#, l
    )).unwrap_or_default();

    let title = doc.meta.title.as_deref().unwrap_or("Presentation");

    format!(include_str!("../../templates/shell.html"),
        title = title,
        ratio = ratio,
        slide_w = dims.0,
        slide_h = dims.1,
        theme_css = theme_css,
        logo_css = logo_css,
        slides = slides_html,
    )
}

fn render_node(node: &Node) -> Option<String> {
    match node {
        Node::Heading(h) => {
            let tag = format!("h{}", h.level);
            let content = render_inline(&h.content);
            Some(format!("<{}>{}</{}>", tag, content, tag))
        }
        Node::Paragraph(inlines) => {
            Some(format!("<p>{}</p>", render_inline(inlines)))
        }
        Node::Quote(inlines) => {
            Some(format!("<div class=\"quote\">{}</div>", render_inline(inlines)))
        }
        Node::Table(t) => {
            let mut s = String::from("<table>\n<thead>\n<tr>");
            for h in &t.headers {
                write!(s, "<th>{}</th>", html_escape(h)).unwrap();
            }
            s.push_str("</tr>\n</thead>\n<tbody>\n");
            for row in &t.rows {
                s.push_str("<tr>");
                for cell in row {
                    write!(s, "<td>{}</td>", html_escape(cell)).unwrap();
                }
                s.push_str("</tr>\n");
            }
            s.push_str("</tbody>\n</table>");
            Some(s)
        }
        Node::Grid(g) => {
            let cols = if g.cols == 0 { g.children.len().max(2) } else { g.cols };
            let mut cls = String::from("grid");
            if !g.class.is_empty() {
                write!(cls, " {}", g.class).unwrap();
            }
            let style = format!("grid-template-columns: repeat({}, 1fr); gap: 16px;", cols);
            let children = g.children.iter()
                .filter_map(|n| render_node(n))
                .collect::<Vec<_>>()
                .join("\n");
            Some(format!(r#"<div class="{}" style="{}">{}</div>"#, cls, style, children))
        }
        Node::Card(c) => {
            let mut cls = String::from("card");
            if !c.class.is_empty() {
                write!(cls, " {}", c.class).unwrap();
            }
            let mut s = format!("<div class=\"{}\">\n", cls);
            if !c.header.is_empty() {
                write!(s, "<h3>{}</h3>\n", c.header).unwrap();
            }
            for line in &c.body {
                write!(s, "<p>{}</p>\n", line).unwrap();
            }
            s.push_str("</div>");
            Some(s)
        }
        Node::List(items) => {
            let mut s = String::from("<ul>\n");
            for item in items {
                write!(s, "<li>{}</li>\n", item).unwrap();
            }
            s.push_str("</ul>");
            Some(s)
        }
        Node::Attr(a) => {
            match a.typ.as_str() {
                "kicker" => Some(format!("<div class=\"kicker\">{}</div>", a.value)),
                "subtitle" => Some(format!("<p class=\"subtitle\">{}</p>", a.value)),
                "speaker" => {
                    let name = a.attrs.iter().find(|(k, _)| k == "name").map(|(_, v)| v.clone()).unwrap_or_default();
                    let role = a.attrs.iter().find(|(k, _)| k == "role").map(|(_, v)| v.clone()).unwrap_or_default();
                    Some(format!("<div class=\"speaker\">{}<span>{}</span></div>", name, role))
                }
                "tiny" => Some(format!("<p class=\"tiny\">{}</p>", a.value)),
                "muted" => Some(format!("<p class=\"muted\">{}</p>", a.value)),
                _ => Some(format!("<div class=\"{}\">{}</div>", a.typ, a.value)),
            }
        }
    }
}

fn render_inline(nodes: &[Inline]) -> String {
    let mut s = String::new();
    for n in nodes {
        match n {
            Inline::Text(t) => s.push_str(&html_escape(t)),
            Inline::Strong(children) => {
                write!(s, "<strong>{}</strong>", render_inline(children)).unwrap();
            }
            Inline::Code(c) => write!(s, "<code>{}</code>", html_escape(c)).unwrap(),
            Inline::SoftBreak => s.push(' '),
            _ => {}
        }
    }
    s
}

fn html_escape(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
}
